File: Binary-tree/105-buildTree/105.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.inorder_id = dict()
        self.preorder: List[int] = []
        self.inorder: List[int] = []

    def helpBuild(
        self, pre_left: int, pre_right: int, in_left: int, in_right: int
    ) -> TreeNode | None:
        if pre_left > pre_right:
            return None

        root_v = self.preorder[pre_left]
        root_id_inorder = self.inorder_id[root_v]
        left_subtree_size = root_id_inorder - in_left

        root = TreeNode(root_v)
        root.left = self.helpBuild(
            pre_left=pre_left + 1,
            pre_right=pre_left + left_subtree_size,
            in_left=in_left,
            in_right=root_id_inorder - 1,
        )
        root.right = self.helpBuild(
            pre_left=pre_left + left_subtree_size + 1,
            pre_right=pre_right,
            in_left=root_id_inorder + 1,
            in_right=in_right,
        )

        return root

    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.preorder = preorder
        self.inorder = inorder
        for id, v in enumerate(inorder):
            self.inorder_id[v] = id
        return self.helpBuild(0, len(inorder) - 1, 0, len(inorder) - 1)

File: Binary-tree/105-buildTree/105.cpp
#include <unordered_map>
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  unordered_map<int, int> inorder_id;  // val -> index (in inorder)

  TreeNode* helpBuild(const vector<int>& preorder, const vector<int>& inorder,
                      int pre_left, int pre_right, int in_left, int in_right) {
    if (pre_left > pre_right) {
      return nullptr;
    }

    int root_v = preorder[pre_left];
    int root_id_inorder = inorder_id[root_v];
    TreeNode* root = new TreeNode(root_v);

    int left_subtree_size = root_id_inorder - in_left;
    root->left =
        helpBuild(preorder, inorder, pre_left + 1, pre_left + left_subtree_size,
                  in_left, root_id_inorder - 1);
    root->right = helpBuild(preorder, inorder, pre_left + left_subtree_size + 1,
                            pre_right, root_id_inorder + 1, in_right);
    return root;
  }

  TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
    int n = inorder.size();
    for (int i = 0; i < n; ++i) {
      inorder_id[inorder[i]] = i;
    }
    return helpBuild(preorder, inorder, 0, n - 1, 0, n - 1);
  }
};